osu.Game/Scoring/Legacy/LegacyScoreEncoder.cs

```diff
@@ -144,17 +149,23 @@ private string replayStringContent
                 // As this is baked into hitobject timing (see `LegacyBeatmapDecoder`) we also need to apply this to replay frame timing.
                 double offset = beatmap?.BeatmapVersion < 5 ? -LegacyBeatmapDecoder.EARLY_VERSION_TIMING_OFFSET : 0;
 
-                int lastTime = 0;
+                double lastTime = 0;
 
                 if (score.Replay != null)
                 {
                     foreach (var f in score.Replay.Frames)
                     {
                         var legacyFrame = getLegacyFrame(f);
 
-                        // Rounding because stable could only parse integral values
-                        int time = (int)Math.Round(legacyFrame.Time + offset);
-                        replayData.Append(FormattableString.Invariant($"{time - lastTime}|{legacyFrame.MouseX ?? 0}|{legacyFrame.MouseY ?? 0}|{(int)legacyFrame.ButtonState},"));
+                        double time = legacyFrame.Time + offset;
+                        // attempt to store a frame delta to the replay with reasonable accuracy.
+                        // precision is constrained to attempt to circumvent potential round-trip parsing issues regarding treatment of floating point numbers.
+                        // 4 decimal places of precision were chosen because in most rulesets the hit window as a function of OD has at most 2 decimal places,
+                        // so adding another 2 decimal places on top should hopefully be a comfortable margin to prevent most rounding issues.
+                        // note that there are still possible sources of inaccuracy here:
+                        // - catastrophic cancellation from the `time - lastTime` operation (probably relevant when timestamps get large in magnitude, i.e. marathon maps)
+                        // - rounding error from rounding the delta time to 4dp
+                        replayData.Append(FormattableString.Invariant($"{Math.Round(time - lastTime, 4)}|{legacyFrame.MouseX ?? 0}|{legacyFrame.MouseY ?? 0}|{(int)legacyFrame.ButtonState},"));
```

Comment:
Would you be willing to reduce this to two decimals? Four seems more than we would ever need, unless I'm missing something in the explanation and this actually helps guarantee avoiding issues.

Reply:
Yeah it's a definite concern. But it would also potentially carve a path towards guaranteed replay playback stability (because regardless of how many DP we store, there's still a chance of a slider path rounding discrepancy).

I guess this can happen as a follow-up, and using 4 D.P. for now is fine though.